src/main/java/build/buildfarm/cas/cfc/CASAccessMetricsRecorder.java

```diff
@@ -0,0 +1,184 @@
+    this.backplane = backplane;
+    this.casEntryReadCountWindow = casEntryReadCountWindow;
+    this.casEntryReadCountUpdateInterval = casEntryReadCountUpdateInterval;
+    this.readIntervalCountQueue = new LinkedList<>();
+    this.queueLength = getQueueLength();
+  }
+
+  /**
+   * Increments the read count by 1 for the given digest.
+   *
+   * @param digest Digest to which the read count is recorded.
+   */
+  public void recordRead(Digest digest) {
+    if (!running) {
+      throw new IllegalStateException("Metrics Recorder is not running");
+    }
+    lock.readLock().lock();
+    try {
+      currentIntervalReadCount.computeIfAbsent(digest, d -> new AtomicInteger(0)).incrementAndGet();
+    } finally {
+      lock.readLock().unlock();
+    }
+  }
+
+  /**
+   * Records the write operation for a digest if it is new. If a read is recorded before for a newly
+   * written digest, this method will have no effect.
+   *
+   * @param digest Digest for which the write operation is recorded.
+   */
+  public void recordWrite(Digest digest) {
+    if (!running) {
+      throw new IllegalStateException("Metrics Recorder is not running");
+    }
+    lock.readLock().lock();
+    try {
+      currentIntervalReadCount.putIfAbsent(digest, new AtomicInteger(0));
+    } finally {
+      lock.readLock().unlock();
+    }
+  }
+
+  /**
+   * Transitions the metrics recorder to the running state and sets up periodic read updates. Once
+   * the metrics recorder is in the running state, calling this method will have no effect.
+   */
+  public synchronized void start() {
+    if (running) {
+      return;
+    }
+    currentIntervalReadCount = new ConcurrentHashMap<>();
+    long initialDelay = getInitialDelayInMillis();
+    scheduler.scheduleAtFixedRate(
+        this::updateReadCount,
+        initialDelay,
+        casEntryReadCountUpdateInterval.toMillis(),
+        TimeUnit.MILLISECONDS);
+    running = true;
+  }
+
+  /** Stops the scheduler responsible to update read counts periodically. */
+  public void stop() throws InterruptedException {
+    running = false;
+    scheduler.shutdown();
+    if (!scheduler.awaitTermination(10, TimeUnit.SECONDS)) {
+      log.severe("Could not shut down cas access metrics recorder service");
+    }
+    scheduler.shutdownNow();
+  }
+
+  private int getQueueLength() {
+    return (int) (casEntryReadCountWindow.toMillis() / casEntryReadCountUpdateInterval.toMillis());
+  }
+
+  private long getInitialDelayInMillis() {
+    // Ensure synchronization among all workers, regardless of their start time.
+    return casEntryReadCountUpdateInterval.toMillis()
+        - (Instant.now().toEpochMilli() % casEntryReadCountUpdateInterval.toMillis());
+  }
+
+  private void updateReadCount() {
+    Stopwatch stopwatch = Stopwatch.createStarted();
+    Map<Digest, AtomicInteger> lastIntervalReadCount;
+    lock.writeLock().lock(); // Ensure all threads writes to new interval.
```

Comment:
Use try-with-resources for these blocks

Reply:
The ReadWriteLock doesn't implements `AutoCloseable`. so moved the closeable logic into separate method. 
i.e `runWithLock`.